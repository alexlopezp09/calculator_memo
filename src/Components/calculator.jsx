import React, { useEffect, useState } from "react";
import { Box } from "@mui/system";
import { Container } from "@mui/material";
import KeyboardComponent from "./keyboard";

const mathOperations = {
  "/": (x, y) => x / y,
  X: (x, y) => x * y,
  "-": (x, y) => x - y,
  "+": (x, y) => x + y,
  "%": (x) => x / 100
};

export const Calculator = () => {
  const [numberContainer, setNumberContainer] = useState([]);
  const [tempNumber, setTempNumber] = useState(null);
  const [operator, setOperator] = useState();
  const [triggerCalculation, setTriggerCalculation] = useState(false);

  console.log("rerendering");
  // Challenge 4 if numbers didn't change and theres no pending
  // operation, avoid recalculating.

  // useEffect(() => {
  //   if (operator === "%") {
  //     setNum(firstNum / 100);
  //   } else if (operator === "+/-") {
  //     if (num > 0) {
  //       setNum(-num);
  //     } else {
  //       setNum(Math.abs(num));
  //     }
  //   }
  // }, [operator, num, firstNum]);

  useEffect(() => {
    // Solo entra al darle click al igual
    if (operator && triggerCalculation) {
      const calculatedValue = mathOperations[operator](
        parseInt(numberContainer[1], 10),
        parseInt(tempNumber, 10)
      );
      setNumberContainer([null, null, calculatedValue]);
      setTriggerCalculation(false);
      setOperator(null);
      setTempNumber(null);
    } else if (operator && tempNumber && !numberContainer[1]) {
      setNumberContainer([null, tempNumber, null]);
      setTempNumber(null);
      if (operator === "%") {
        setTriggerCalculation(true);
      }
    }
  }, [triggerCalculation, operator, numberContainer, tempNumber]);

  const changeNumber = (e) => {
    let input = e.target.value;
    setTempNumber((prev) => (prev ? prev + input : input));
  };

  // function changeSign() {
  //   if (num > 0) {
  //     setNum(-num);
  //   } else {
  //     setNum(Math.abs(num));
  //   }
  // }

  const clearScreen = () => {
    setNumberContainer([]);
    setTempNumber(null);
    setOperator(null);
  };

  return (
    <div>
      <Box m={5} />
      <Container maxWidth="xs">
        <div className="wrapper">
          <Box m={12} />
          <h1 className="resultado">{tempNumber || numberContainer[2] || 0}</h1>
          <KeyboardComponent
            setOperator={setOperator}
            clearScreen={clearScreen}
            changeNumber={changeNumber}
            triggerCalculation={setTriggerCalculation}
          />
        </div>
      </Container>
    </div>
  );
};
