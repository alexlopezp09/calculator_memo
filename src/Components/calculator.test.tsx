import { render } from "@testing-library/react";
import { Calculator } from "./calculator";
import { expect, test } from "@jest/globals";

test("render footer module", async () => {
  const { container } = render(<Calculator />);
  expect(container).toMatchSnapshot();
});
