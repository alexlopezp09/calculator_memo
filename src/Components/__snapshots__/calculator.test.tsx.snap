// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render footer module 1`] = `
<div>
  <div>
    <div
      class="MuiBox-root css-1qapnz0"
    />
    <div
      class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
    >
      <div
        class="wrapper"
      >
        <div
          class="MuiBox-root css-1r9xrk5"
        />
        <h1
          class="resultado"
        >
          0
        </h1>
        <button
          value="AC"
        >
          AC
        </button>
        <button
          value="+/-"
        >
          +/-
        </button>
        <button
          value="%"
        >
          %
        </button>
        <button
          class="orange"
          value="/"
        >
          /
        </button>
        <button
          class="grey"
          value="7"
        >
          7
        </button>
        <button
          class="grey"
          value="8"
        >
          8
        </button>
        <button
          class="grey"
          value="9"
        >
          9
        </button>
        <button
          class="orange"
          value="X"
        >
          X
        </button>
        <button
          class="grey"
          value="4"
        >
          4
        </button>
        <button
          class="grey"
          value="5"
        >
          5
        </button>
        <button
          class="grey"
          value="6"
        >
          6
        </button>
        <button
          class="orange"
          value="-"
        >
          -
        </button>
        <button
          class="grey"
          value="1"
        >
          1
        </button>
        <button
          class="grey"
          value="2"
        >
          2
        </button>
        <button
          class="grey"
          value="3"
        >
          3
        </button>
        <button
          class="orange"
          value="+"
        >
          +
        </button>
        <button
          class="grey"
          value="0"
        >
          0
        </button>
        <button
          style="visibility: hidden;"
        >
          k
        </button>
         
        <button
          class="grey"
          value="."
        >
          ,
        </button>
        <button
          class="orange"
          value="="
        >
          =
        </button>
      </div>
    </div>
  </div>
</div>
`;
